File: contests/lc/1339.MaximumProductofSplittedBinaryTree/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val: int = 0, left: Optional["TreeNode"] = None, right: Optional["TreeNode"] = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        sums = []
        total = self._subtree_sum(root, sums)
        best = 0
        for s in sums:
            product = s * (total - s)
            if product > best:
                best = product
        return best % 1_000_000_007

    def _subtree_sum(self, node: Optional[TreeNode], sums: list[int]) -> int:
        if node is None:
            return 0

        left = self._subtree_sum(node.left, sums)
        right = self._subtree_sum(node.right, sums)
        current = left + right + node.val
        sums.append(current)
        return current

File: contests/lc/1339.MaximumProductofSplittedBinaryTree/solution.go
package main

const mod int64 = 1_000_000_007

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxProduct(root *TreeNode) int {
	if root == nil {
		return 0
	}

	sums := make([]int64, 0, 1)
	total := subtreeSum(root, &sums)

	var best int64
	for _, s := range sums {
		product := s * (total - s)
		if product > best {
			best = product
		}
	}

	return int(best % mod)
}

func subtreeSum(node *TreeNode, sums *[]int64) int64 {
	if node == nil {
		return 0
	}

	left := subtreeSum(node.Left, sums)
	right := subtreeSum(node.Right, sums)
	current := left + right + int64(node.Val)
	*sums = append(*sums, current)
	return current
}
